torch/csrc/jit/codegen/cuda/fusion_segmenter.cpp

```diff
@@ -0,0 +1,916 @@
+// destruction will return fusion to original state.
+// Not used temporarily but will be useful when adding more mergin heuristics
+class FusionSegmentGuard {
+ public:
+  FusionSegmentGuard() = delete;
+  FusionSegmentGuard(FusionSegmentGuard const&) = delete;
+  FusionSegmentGuard& operator=(FusionSegmentGuard const&) = delete;
+
+  FusionSegmentGuard(
+      Fusion* fusion,
+      std::vector<Val*> inputs,
+      std::vector<Val*> outputs)
+      : fusion_(fusion),
+        old_inputs(fusion->inputs()),
+        old_outputs(fusion->outputs()),
+        new_inputs(std::move(inputs)),
+        new_outputs(std::move(outputs)) {
+    for (auto old_inp : old_inputs) {
+      fusion_->removeInput(old_inp);
+    }
+
+    for (auto old_out : old_outputs) {
+      fusion_->removeOutput(old_out);
+    }
+
+    for (auto new_inp : new_inputs) {
+      fusion_->addInput(new_inp);
+    }
+
+    for (auto new_out : new_outputs) {
+      fusion_->addOutput(new_out);
+    }
+  }
+
+  ~FusionSegmentGuard() {
+    for (auto new_inp : new_inputs) {
+      fusion_->removeInput(new_inp);
+    }
+
+    for (auto new_out : new_outputs) {
+      fusion_->removeOutput(new_out);
+    }
+
+    for (auto old_inp : old_inputs) {
+      fusion_->addInput(old_inp);
+    }
+
+    for (auto old_out : old_outputs) {
+      fusion_->addOutput(old_out);
+    }
+  }
+
+ private:
+  Fusion* const fusion_;
+  const std::vector<Val*> old_inputs;
+  const std::vector<Val*> old_outputs;
+  const std::vector<Val*> new_inputs;
+  const std::vector<Val*> new_outputs;
+};
+
+c10::optional<ScheduleHeuristic> tryMerge(
+    Fusion* fusion,
+    SegmentedGroup* a,
+    SegmentedGroup* b = nullptr) {
+  FusionSegmentGuard fsg(fusion, getAllInputs(a, b), getAllOutputs(a, b));
+
+  return SchedulerEntry::proposeHeuristics(fusion);
+}
+
+} // namespace
+
+bool SegmentCandidateFinder::codeGenSupportedMerge(SegmentedEdge* edge) {
+  Fusion* fusion = &segmented_fusion->completeFusion();
+  auto h = tryMerge(fusion, edge->from_, edge->to_);
+  return h.has_value();
+}
+
+// TODO: consider caching the heuristics value so tryMerge doesn't have to be
+//       called twice
+ScheduleHeuristic SegmentCandidateFinder::deriveHeuristic(
+    SegmentedGroup* group) {
+  Fusion* fusion = &segmented_fusion->completeFusion();
+  auto h = tryMerge(fusion, group);
+  TORCH_INTERNAL_ASSERT(h.has_value());
+  return h.value();
+}
+
+SegmentCandidateFinder::SegmentCandidateFinder(const Fusion* fusion) {
+  segmented_fusion = std::make_unique<SegmentedFusion>(fusion);
+  findSegments();
+}
+
+void SegmentCandidateFinder::findSegments() {
+  // TODO: Make traversal items local to this function.
+
+  // Need this for initialization of the DAG that is process
+  std::unordered_map<Expr*, SegmentedGroup*> expr2group;
+
+  // Initialize DAG, convert each expr to a segment group
+  size_t total_exprs = 0;
+  for (auto expr : completeFusion().exprs()) {
```

Comment:
Might want to store `completeFusion().exprs()` in a variable so that it can be reused below. Shouldn't be a performance bottleneck, but `Fusion::exprs()` involves a graph traversal.

Reply:
Nice catch! Updated, thanks for pointing this out. 👍